File: Week-Wise-Assignments/Assignment 4/Compressed Solutions/C Programs/2_merge_sorted.c
#include <stdio.h>
#include <stdlib.h>

struct Node {
    int data;
    struct Node* next;
};

struct Node* createNode(int data) {
    struct Node* node = malloc(sizeof(struct Node));
    node->data = data;
    node->next = NULL;
    return node;
}

struct Node* merge(struct Node* l1, struct Node* l2) {
    if (!l1) return l2;
    if (!l2) return l1;
    
    if (l1->data <= l2->data) {
        l1->next = merge(l1->next, l2);
        return l1;
    } else {
        l2->next = merge(l1, l2->next);
        return l2;
    }
}

void display(struct Node* head) {
    while (head) {
        printf("%d ", head->data);
        head = head->next;
    }
    printf("\n");
}

int main() {
    struct Node* l1 = createNode(1);
    l1->next = createNode(3);
    l1->next->next = createNode(5);
    
    struct Node* l2 = createNode(2);
    l2->next = createNode(4);
    l2->next->next = createNode(6);
    
    printf("List 1: ");
    display(l1);
    printf("List 2: ");
    display(l2);
    
    struct Node* merged = merge(l1, l2);
    printf("Merged: ");
    display(merged);
    
    return 0;
}

File: Week-Wise-Assignments/Assignment 4/Compressed Solutions/Python Programs/2_merge_sorted.py
class Node:
    def __init__(self, data):
        self.data = data
        self.next = None

def merge_recursive(l1, l2):
    if not l1: return l2
    if not l2: return l1
    
    if l1.data <= l2.data:
        l1.next = merge_recursive(l1.next, l2)
        return l1
    else:
        l2.next = merge_recursive(l1, l2.next)
        return l2

def display(head):
    while head:
        print(head.data, end=" -> " if head.next else " -> NULL\n")
        head = head.next

# Test
l1 = Node(1)
l1.next = Node(3)
l1.next.next = Node(5)

l2 = Node(2)
l2.next = Node(4)
l2.next.next = Node(6)

print("List 1: ", end="")
display(l1)
print("List 2: ", end="")
display(l2)

merged = merge_recursive(l1, l2)
print("Merged: ", end="")
display(merged)